### PR TITLE
feat: enable L4 stream logging

### LINKED ISSUE #5400: request help: How to enable logging for the four-layer proxy stream route?
### Issue description

七层的http日志是存储在acccess.log中的

四层代理stream route如何启用日志并自定义日志路径呢？

传统的openresty是可以实现的

### Environment

最新版

### PR DESCRIPTION
**Pull request by bisakhmondal**
### What this PR does / why we need it:

closes #5400 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: conf/config-default.yaml [hunk 1]
```diff
@@ -174,6 +174,11 @@ nginx_config:                     # config for render the template to generate n
   #  - TEST_ENV
 
   stream:
+    enable_access_log: true         # enable access log or not, default true
+    access_log: logs/access_stream.log
+    access_log_format: "$remote_addr [$time_local] $status $bytes_sent $session_time $upstream_connect_time \"$protocol://$upstream_addr\""
```
Comment (bisakhmondal) [inline conf/config-default.yaml:179]: @spacewander by consulting http://nginx.org/en/docs/varindex.html, I created the log format. Do you think we can do better? Do let me know. Thanks.

File: apisix/cli/ngx_tpl.lua [hunk 2]
```diff
@@ -84,6 +84,16 @@ stream {
...
 
+    # for stream logs
+    {% if stream.enable_access_log == false then %}
+    access_log off;
+    {% else %}
+    log_format main escape={* stream.access_log_format_escape *} '{* stream.access_log_format *}';
+
+    access_log {* stream.access_log *} main buffer=16384 flush=3;
+    {% end %}
+    open_log_file_cache max=1000 inactive=60;
```
Comment (spacewander) [inline apisix/cli/ngx_tpl.lua:95]: We don't need it, as the log file name doesn't contain variable. There is only one open log file.
See https://nginx.org/en/docs/http/ngx_http_log_module.html#open_log_file_cache

File: apisix/cli/ngx_tpl.lua [hunk 3]
```diff
@@ -261,7 +271,7 @@ http {
 
     access_log {* http.access_log *} main buffer=16384 flush=3;
     {% end %}
-    open_file_cache  max=1000 inactive=60;
```
Comment (spacewander) [inline apisix/cli/ngx_tpl.lua:264]: open_file_cache is there: https://nginx.org/en/docs/http/ngx_http_core_module.html#open_file_cache
This directive is different from open_log_file_cache.

File: conf/config-default.yaml [hunk 4]
```diff
@@ -174,6 +174,11 @@ nginx_config:                     # config for render the template to generate n
   #  - TEST_ENV
 
   stream:
+    enable_access_log: true         # enable access log or not, default true
```
Comment (spacewander) [inline conf/config-default.yaml:177]: The access_log in stream is false by default: https://nginx.org/en/docs/stream/ngx_stream_log_module.html#access_log

File: conf/config-default.yaml (hunk 1 above)
Comment (spacewander) [reply conf/config-default.yaml:179]: I think we can use Nginx's basic format in its example?
```
log_format basic '$remote_addr [$time_local] '
                 '$protocol $status $bytes_sent $bytes_received '
                 '$session_time';
```
See https://nginx.org/en/docs/stream/ngx_stream_log_module.html.

Kong also uses it too: https://github.com/Kong/kong/blob/d65101fe80fd7ac9870a84d34d81bda8bcb461ac/kong/templates/nginx_kong_stream.lua#L3

**Review by spacewander — COMMENTED** [review]
Could you add a test in https://github.com/apache/apisix/blob/master/t/cli/test_access_log.sh? Thanks.

File: conf/config-default.yaml (hunk 1 above)
Comment (bisakhmondal) [reply conf/config-default.yaml:179]: Okay.

File: conf/config-default.yaml (hunk 4 above)
Comment (bisakhmondal) [reply conf/config-default.yaml]: fixed

File: apisix/cli/ngx_tpl.lua (hunk 3 above)
Comment (bisakhmondal) [reply apisix/cli/ngx_tpl.lua]: Ack

File: apisix/cli/ngx_tpl.lua (hunk 2 above)
Comment (bisakhmondal) [reply apisix/cli/ngx_tpl.lua]: Ack. Thanks

**Comment by bisakhmondal** [comment]
> Could you add a test in https://github.com/apache/apisix/blob/master/t/cli/test_access_log.sh? Thanks.

Done. Took a while to figure out how to set up mock upstream.

File: t/cli/test_access_log.sh [hunk 5]
```diff
@@ -224,3 +224,41 @@ fi
...
+sleep 4
+tail -n 1 logs/access_stream.log > output.log
+
+if ! grep '127.0.0.1 UDP test_stream_access_log_format' output.log; then
+    echo "failed: should have found udp log entry"
+    cat output.log
+    exit 1
+fi
+
+make stop
```
Comment (spacewander) [inline t/cli/test_access_log.sh:263]: We will clean up automatically, so there is no need to add it. See https://github.com/apache/apisix/blob/02d32c9dbc0813e68f51075b2ff8a067f60fca6a/t/cli/common.sh#L32

Comment (bisakhmondal) [reply t/cli/test_access_log.sh]: Awesome. Thanks for sharing the details. Updated.

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**